array/array.go

```diff
@@ -0,0 +1,131 @@
+package array
+
+import (
+	"errors"
+)
+
+type Array struct {
+	arr  []int
+	size int
+}
+
+func NewArray(capacity int) *Array {
+	return &Array{
+		arr: make([]int, capacity),
+	}
+}
+
+func (a *Array) Size() int {
+	return a.size
+}
+
+func (a *Array) Capacity() int {
+	return len(a.arr)
+}
+
+func (a *Array) IsEmpty() bool {
+	return a.Size() == 0
+}
+
+func (a *Array) ValueAt(index int) (int, error) {
+	if index < 0 || index >= a.Size() {
+		return -1, errors.New("Index is out of range")
+	}
+	return a.arr[index], nil
+}
+
+func (a *Array) Insert(item int, index int) error {
+	if index < 0 || index >= a.Capacity() {
```

Comment:
Valid range is from `0` to `a.size` so `index >= a.Capacity()` must be `index >= a.size`

Reply:
The intention of this function is to allow inserting at any index in range 0 to capacity - 1, which mimics the behavior in JavaScript.

For example, in JS I can do:
```javascript
const arr = []
arr[6] = 1
console.log(arr.length) // output: 7
```
Every other elements from index 0 to 5 are now `undefined`.

So in my implementation, if I do:
```go
arr := NewArray(10)
arr.Insert(1, 6)
fmt.Print(arr.Size()) // output: 7
```
Every other elements from index 0 to 5 would be `0`, since Go `int` type can't be `nil`.

Reply:
Actually JS arrays don't have a concept of size but we have. We implement array to understand how an array works in most languages and JS is an edge case. Inserting at an index which is greater than size is an unexpected behavior. We don't know which value will be automatically inserted. Normally, calling `insert` once will only add a value into array but in this case, `size` can be increased from 0 to 7 after an insertion.

@thanks173 How do you think?

Reply:
I understand the purpose of learning to understand how array works. My implementation looks like a wrapper class with enhance functionality, I guess. :)

Reply:
In my point of view, JS is a mess. We should not use it as a best practice.